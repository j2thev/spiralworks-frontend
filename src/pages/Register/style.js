import styled from 'styled-components';
import { Container } from 'reactstrap';

export const RegisterContainer = styled(Container)`
  margin-top: 50px
`;
