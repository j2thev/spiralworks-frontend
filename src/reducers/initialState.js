export default {
  user: {
    id: '',
    email: '',
    phoneNo: '',
    firstName: '',
    lastName: '',
    isAuthenticated: false
  }
}
