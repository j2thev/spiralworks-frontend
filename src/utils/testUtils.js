import React from 'react';
import { Provider } from 'react-redux';

export function TestProvider({ children, store }) {
  return <Provider store={store}>{children}</Provider>;
}
