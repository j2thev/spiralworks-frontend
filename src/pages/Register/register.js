import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import {
  Row,
  Col,
  Form,
  FormGroup,
  Label,
  Input,
  Button,
  FormText,
  Alert
} from 'reactstrap';
import * as R from './style';
import { map, has } from 'lodash';
import * as userActions from '../../actions/userActions';

class Register extends Component {
  constructor(props) {
    super(props);

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);

    this.state = {
      email: '',
      phoneNo: '',
      firstName: '',
      lastName: '',
      password: '',
      confirmPassword: '',
      errors: {
        email: '',
        phoneNo: '',
        firstName: '',
        lastName: '',
        password: '',
        confirmPassword: ''
      },
      alert: {
        color: '',
        message: '',
        visible: false
      }
    }
  }

  handleChange(event) {
    const { name, value } = event.target;
    const { errors, password } = this.state;

    switch (name) {
      case 'email':
      case 'phoneNo': 
      case 'firstName':
      case 'lastName':
      case 'password': {
        Object.assign(errors, { [name]: '' });
        break;
      }
      case 'confirmPassword': {
        Object.assign(errors, {
          confirmPassword: (password === value) ? '' : `Password doesn't match`
        });
        break;
      }
      default: {
        break;
      }
    }

    this.setState({
      [name]: value,
      errors
    });
  }

  validateForm(errors) {
    let valid = true;

    map(errors, (error) => error.length > 0 && (valid = false));

    return valid;
  }

  onShowAlert(alert) {
    this.setState({ alert }, () => {
      setTimeout(() => {
        this.setState({
          alert: {
            visible: false
          }
        });
      }, 5000);
    });
  }

  handleSubmit(event) {
    event.preventDefault();

    const { create } = this.props;
    const { errors } = this.state;

    map(event.target, (node) => {
      const { name, validationMessage } = node;
      const value = (name === 'confirmPassword') ? (errors.confirmPassword || validationMessage) : validationMessage;

      Object.assign(errors, { [name]: value });
    });

    if (this.validateForm(errors)) {
      const { email, phoneNo, firstName, lastName, password } = this.state;
      const user = { email, phoneNo, firstName, lastName, password };

      create(user)
        .then(result => {
          const alert = {
            color: 'success',
            message: 'Your account has been successfully created!',
            visible: true
          };

          this.onShowAlert(alert);
        })
        .catch(error => {
          const alert = {
            color: 'danger',
            visible: true
          };

          if (has(error, 'response.data.error')) {
            const { errors } = error.response.data.error;
            map(errors, (field) => {
              Object.assign(alert, { message: field.message });
            });
          } else {
            Object.assign(alert, { message: error.message });
          }
          
          this.onShowAlert(alert);
        });
    } else {
      this.setState({ errors });
    }
  }

  render() {
    const { email, phoneNo, firstName, lastName, password, confirmPassword, errors, alert } = this.state;

    return (
      <R.RegisterContainer>
        <Row>
          <Col sm="12" md={{ size: 6, offset: 3 }}>
            <Form onSubmit={this.handleSubmit} noValidate>
              <FormGroup>
                <h4>Create your personal account</h4>
              </FormGroup>
              <FormGroup>
                <Alert color={alert.color} isOpen={alert.visible}>
                  {alert.message}
                </Alert>
              </FormGroup>
              <FormGroup>
                <Label>First Name</Label>
                <Input
                  invalid={errors.firstName.length > 0 ? true : false}
                  type="text"
                  value={firstName}
                  required
                  name="firstName"
                  onChange={this.handleChange}
                />
                <FormText color="danger">
                  {errors.firstName}
                </FormText>
              </FormGroup>
              <FormGroup>
                <Label>Last Name</Label>
                <Input
                  invalid={errors.lastName.length > 0 ? true : false}
                  type="text"
                  value={lastName}
                  required
                  name="lastName"
                  onChange={this.handleChange}
                />
                <FormText color="danger">
                  {errors.lastName}
                </FormText>
              </FormGroup>
              <FormGroup>
                <Label>Email</Label>
                <Input
                  invalid={errors.email.length > 0 ? true : false}
                  type="email"
                  value={email}
                  required
                  name="email"
                  onChange={this.handleChange}
                  error={errors.email} />
                <FormText color="danger">
                  {errors.email}
                </FormText>
              </FormGroup>
              <FormGroup>
                <Label>Phone Number (Format: +639 or 09 xxxxxxxxx)</Label>
                <Input
                  invalid={errors.phoneNo.length > 0 ? true : false}
                  type="phoneNo"
                  value={phoneNo}
                  required
                  pattern="^(09|\+639)\d{9}$"
                  name="phoneNo"
                  onChange={this.handleChange}
                  error={errors.phoneNo}
                />
                <FormText color="danger">
                  {errors.phoneNo}
                </FormText>
              </FormGroup>
              <FormGroup>
                <Label>Password</Label>
                <Input
                  invalid={errors.password.length > 0 ? true : false}
                  type="password"
                  value={password}
                  required
                  name="password"
                  onChange={this.handleChange}
                />
                <FormText color="danger">
                  {errors.password}
                </FormText>
              </FormGroup>
              <FormGroup>
                <Label>Confirm Password</Label>
                <Input
                  invalid={errors.confirmPassword.length > 0 ? true : false}
                  type="password"
                  value={confirmPassword}
                  required
                  name="confirmPassword"
                  onChange={this.handleChange}
                />
                <FormText color="danger">
                  {errors.confirmPassword}
                </FormText>
              </FormGroup>
              <Button color="success" block>Create an account</Button>
            </Form>
          </Col>
        </Row>
      </R.RegisterContainer>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return bindActionCreators(userActions, dispatch);
};

export default connect(null, mapDispatchToProps)(Register);
