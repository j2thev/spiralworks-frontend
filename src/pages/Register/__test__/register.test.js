import React from 'react';
import configuredMockStore from 'redux-mock-store';
import toJSON from 'enzyme-to-json';
import { mount } from 'enzyme';
import { TestProvider } from '../../../utils/testUtils';
import initialState from '../../../reducers/initialState';
import '../../../tools/setupEnzyme';
import Register from '../../Register';

describe('Register Page', () => {
  it('should render correctly', () => {
    const store = configuredMockStore()(initialState);
    const container = mount(
      <TestProvider store={store}>
        <Register />
      </TestProvider>
    );

    expect(toJSON(container)).toMatchSnapshot();
  });
});
