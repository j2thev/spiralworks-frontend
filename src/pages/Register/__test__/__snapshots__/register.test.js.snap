// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Page should render correctly 1`] = `
<TestProvider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Provider
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <Connect(Register)>
      <Register
        create={[Function]}
        login={[Function]}
        logout={[Function]}
        save={[Function]}
        update={[Function]}
      >
        <Styled(Container)>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bdVaJa",
                  "isStatic": true,
                  "lastClassName": "gOVJvL",
                  "rules": Array [
                    "
  margin-top: 50px
",
                  ],
                },
                "displayName": "Styled(Container)",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bdVaJa",
                "target": [Function],
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <Container
              className="sc-bdVaJa gOVJvL"
              tag="div"
            >
              <div
                className="sc-bdVaJa gOVJvL container"
              >
                <Row
                  tag="div"
                >
                  <div
                    className="row"
                  >
                    <Col
                      md={
                        Object {
                          "offset": 2,
                          "size": 8,
                        }
                      }
                      sm="12"
                      tag="div"
                      widths={
                        Array [
                          "xs",
                          "sm",
                          "md",
                          "lg",
                          "xl",
                        ]
                      }
                    >
                      <div
                        className="col-sm-12 col-md-8 offset-md-2"
                      >
                        <Form
                          noValidate={true}
                          onSubmit={[Function]}
                          tag="form"
                        >
                          <form
                            className=""
                            noValidate={true}
                            onSubmit={[Function]}
                          >
                            <FormGroup
                              tag="div"
                            >
                              <div
                                className="form-group"
                              >
                                <h3>
                                  Create your personal account
                                </h3>
                              </div>
                            </FormGroup>
                            <FormGroup
                              tag="div"
                            >
                              <div
                                className="form-group"
                              >
                                <Alert
                                  closeAriaLabel="Close"
                                  color=""
                                  fade={true}
                                  isOpen={false}
                                  tag="div"
                                  transition={
                                    Object {
                                      "appear": true,
                                      "baseClass": "fade",
                                      "baseClassActive": "show",
                                      "enter": true,
                                      "exit": true,
                                      "in": true,
                                      "mountOnEnter": false,
                                      "onEnter": [Function],
                                      "onEntered": [Function],
                                      "onEntering": [Function],
                                      "onExit": [Function],
                                      "onExited": [Function],
                                      "onExiting": [Function],
                                      "tag": "div",
                                      "timeout": 150,
                                      "unmountOnExit": true,
                                    }
                                  }
                                >
                                  <Fade
                                    appear={true}
                                    baseClass="fade"
                                    baseClassActive="show"
                                    className="alert alert-"
                                    enter={true}
                                    exit={true}
                                    in={false}
                                    mountOnEnter={false}
                                    onEnter={[Function]}
                                    onEntered={[Function]}
                                    onEntering={[Function]}
                                    onExit={[Function]}
                                    onExited={[Function]}
                                    onExiting={[Function]}
                                    role="alert"
                                    tag="div"
                                    timeout={150}
                                    unmountOnExit={true}
                                  >
                                    <Transition
                                      appear={true}
                                      enter={true}
                                      exit={true}
                                      in={false}
                                      mountOnEnter={false}
                                      onEnter={[Function]}
                                      onEntered={[Function]}
                                      onEntering={[Function]}
                                      onExit={[Function]}
                                      onExited={[Function]}
                                      onExiting={[Function]}
                                      timeout={150}
                                      unmountOnExit={true}
                                    />
                                  </Fade>
                                </Alert>
                              </div>
                            </FormGroup>
                            <FormGroup
                              tag="div"
                            >
                              <div
                                className="form-group"
                              >
                                <Label
                                  tag="label"
                                  widths={
                                    Array [
                                      "xs",
                                      "sm",
                                      "md",
                                      "lg",
                                      "xl",
                                    ]
                                  }
                                >
                                  <label
                                    className=""
                                  >
                                    First Name
                                  </label>
                                </Label>
                                <Input
                                  invalid={false}
                                  name="firstName"
                                  onChange={[Function]}
                                  required={true}
                                  type="text"
                                  value=""
                                >
                                  <input
                                    className="form-control"
                                    name="firstName"
                                    onChange={[Function]}
                                    required={true}
                                    type="text"
                                    value=""
                                  />
                                </Input>
                                <FormText
                                  color="danger"
                                  tag="small"
                                >
                                  <small
                                    className="form-text text-danger"
                                  />
                                </FormText>
                              </div>
                            </FormGroup>
                            <FormGroup
                              tag="div"
                            >
                              <div
                                className="form-group"
                              >
                                <Label
                                  tag="label"
                                  widths={
                                    Array [
                                      "xs",
                                      "sm",
                                      "md",
                                      "lg",
                                      "xl",
                                    ]
                                  }
                                >
                                  <label
                                    className=""
                                  >
                                    Last Name
                                  </label>
                                </Label>
                                <Input
                                  invalid={false}
                                  name="lastName"
                                  onChange={[Function]}
                                  required={true}
                                  type="text"
                                  value=""
                                >
                                  <input
                                    className="form-control"
                                    name="lastName"
                                    onChange={[Function]}
                                    required={true}
                                    type="text"
                                    value=""
                                  />
                                </Input>
                                <FormText
                                  color="danger"
                                  tag="small"
                                >
                                  <small
                                    className="form-text text-danger"
                                  />
                                </FormText>
                              </div>
                            </FormGroup>
                            <FormGroup
                              tag="div"
                            >
                              <div
                                className="form-group"
                              >
                                <Label
                                  tag="label"
                                  widths={
                                    Array [
                                      "xs",
                                      "sm",
                                      "md",
                                      "lg",
                                      "xl",
                                    ]
                                  }
                                >
                                  <label
                                    className=""
                                  >
                                    Email
                                  </label>
                                </Label>
                                <Input
                                  error=""
                                  invalid={false}
                                  name="email"
                                  onChange={[Function]}
                                  required={true}
                                  type="email"
                                  value=""
                                >
                                  <input
                                    className="form-control"
                                    error=""
                                    name="email"
                                    onChange={[Function]}
                                    required={true}
                                    type="email"
                                    value=""
                                  />
                                </Input>
                                <FormText
                                  color="danger"
                                  tag="small"
                                >
                                  <small
                                    className="form-text text-danger"
                                  />
                                </FormText>
                              </div>
                            </FormGroup>
                            <FormGroup
                              tag="div"
                            >
                              <div
                                className="form-group"
                              >
                                <Label
                                  tag="label"
                                  widths={
                                    Array [
                                      "xs",
                                      "sm",
                                      "md",
                                      "lg",
                                      "xl",
                                    ]
                                  }
                                >
                                  <label
                                    className=""
                                  >
                                    Phone Number
                                  </label>
                                </Label>
                                <Input
                                  error=""
                                  invalid={false}
                                  name="phoneNo"
                                  onChange={[Function]}
                                  pattern="^(09|\\\\+639)\\\\d{9}$"
                                  required={true}
                                  type="phoneNo"
                                  value=""
                                >
                                  <input
                                    className="form-control"
                                    error=""
                                    name="phoneNo"
                                    onChange={[Function]}
                                    pattern="^(09|\\\\+639)\\\\d{9}$"
                                    required={true}
                                    type="phoneNo"
                                    value=""
                                  />
                                </Input>
                                <FormText
                                  color="muted"
                                  tag="small"
                                >
                                  <small
                                    className="form-text text-muted"
                                  >
                                    Format: +639xxxxxxxxx, 09xxxxxxxxx
                                  </small>
                                </FormText>
                                <FormText
                                  color="danger"
                                  tag="small"
                                >
                                  <small
                                    className="form-text text-danger"
                                  />
                                </FormText>
                              </div>
                            </FormGroup>
                            <FormGroup
                              tag="div"
                            >
                              <div
                                className="form-group"
                              >
                                <Label
                                  tag="label"
                                  widths={
                                    Array [
                                      "xs",
                                      "sm",
                                      "md",
                                      "lg",
                                      "xl",
                                    ]
                                  }
                                >
                                  <label
                                    className=""
                                  >
                                    Password
                                  </label>
                                </Label>
                                <Input
                                  invalid={false}
                                  name="password"
                                  onChange={[Function]}
                                  required={true}
                                  type="password"
                                  value=""
                                >
                                  <input
                                    className="form-control"
                                    name="password"
                                    onChange={[Function]}
                                    required={true}
                                    type="password"
                                    value=""
                                  />
                                </Input>
                                <FormText
                                  color="danger"
                                  tag="small"
                                >
                                  <small
                                    className="form-text text-danger"
                                  />
                                </FormText>
                              </div>
                            </FormGroup>
                            <FormGroup
                              tag="div"
                            >
                              <div
                                className="form-group"
                              >
                                <Label
                                  tag="label"
                                  widths={
                                    Array [
                                      "xs",
                                      "sm",
                                      "md",
                                      "lg",
                                      "xl",
                                    ]
                                  }
                                >
                                  <label
                                    className=""
                                  >
                                    Confirm Password
                                  </label>
                                </Label>
                                <Input
                                  invalid={false}
                                  name="confirmPassword"
                                  onChange={[Function]}
                                  required={true}
                                  type="password"
                                  value=""
                                >
                                  <input
                                    className="form-control"
                                    name="confirmPassword"
                                    onChange={[Function]}
                                    required={true}
                                    type="password"
                                    value=""
                                  />
                                </Input>
                                <FormText
                                  color="danger"
                                  tag="small"
                                >
                                  <small
                                    className="form-text text-danger"
                                  />
                                </FormText>
                              </div>
                            </FormGroup>
                            <Button
                              block={true}
                              color="success"
                              tag="button"
                            >
                              <button
                                aria-label={null}
                                className="btn btn-success btn-block"
                                onClick={[Function]}
                              >
                                CREATE AN ACCOUNT
                              </button>
                            </Button>
                          </form>
                        </Form>
                      </div>
                    </Col>
                  </div>
                </Row>
              </div>
            </Container>
          </StyledComponent>
        </Styled(Container)>
      </Register>
    </Connect(Register)>
  </Provider>
</TestProvider>
`;
